### PR TITLE
Add block variations to the slash inserter

### LINKED ISSUE #20583: Block API: Integrate variations with the `/` inserter
## Description

Follow-up for #16283.

With https://github.com/WordPress/gutenberg/pull/19243, we integrate block variations API with the inserter. We still should seek a way to do a follow-up for the `/` inserter. At the moment, it only lists block types.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #20583

Uses the API introduced in #22853 to introduce support for block variations in the slash inserter.
More importantly, this PR makes the slash inserter behaves exactly like the global inserter by sharing logic.
We could for example easily consider adding patterns in a separate PR.

### CODE REVIEW

File: packages/block-editor/src/components/inserter/utils.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/**
+ * Normalizes an inserter block types list and includes variations as separate items.
+ *
+ * @param {Array} items Denormalized inserter items
+ * @return {Array} Normalized inserter items.
+ */
+export function includeVariationsInInserterItems( items ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/utils.js:7]: I think should probably be bundled in the `getInserterItems` selector. But I'd leave that for a separate PR as this is already how it works today in the global inserter, it's just extracted as a function. cc @gziolo @noisysocks

Comment (gziolo) [reply packages/block-editor/src/components/inserter/utils.js:7]: I need to double-check the initial implementation for block variation tomorrow. I think there was a reason to keep it as is initially. It might be no longer valid concern :)

Comment (gziolo) [reply packages/block-editor/src/components/inserter/utils.js:7]: I found related PR: #19243. I can't think of any blocker for further refactoring. It seems like producing multiple items for blocks with variations followed-up with array flattening would do the trick. We just need to give it a try :)

File: packages/block-editor/src/autocompleters/block.js [hunk 2]
```diff
@@ -1,85 +1,38 @@
...
-/**
- * Returns the name of the currently selected block.
- *
- * @return {string?} The name of the currently selected block or `null` if no
- *                   block is selected.
- */
-function defaultGetSelectedBlockName() {
-	const { getSelectedBlockClientId, getBlockName } = select(
-		'core/block-editor'
+const createBlocksFromInnerBlocksTemplate = ( innerBlocksTemplate ) => {
```
Comment (gziolo) [inline packages/block-editor/src/autocompleters/block.js:21]: It's copied in a few places, we should extract it and put in `@wordpress/blocks` as public API.

Comment (youknowriad) [reply packages/block-editor/src/autocompleters/block.js:21]: Right

Comment (youknowriad) [reply packages/block-editor/src/autocompleters/block.js:21]: Do you know if the "variations" are the only place where we could have this shape of innerBlocks.
It seems to me there's an inconsistency here. The variationIItem has a shape of a block object `{ name, innerBlocks, attributes }` but  its  innerBlocks have  a shape of a template `[ name, attributes, innerBlocks ]` which make me thing we should consolidate or rename `innerBlocks` to `innerBlocksTemplate` here.

Comment (youknowriad) [reply packages/block-editor/src/autocompleters/block.js:21]: I just figured that it's because in the variations API we have innerBlocks property with the template shape too. I think that's not great as it's inconsistent with "example" API and it's inconsistent with it's self since a variation is an object with attributes and innerBlocks so you'd assume its innerBlocks keep the same shape.

I'll address this in a separate PR and I'll drop the template format from the variations while retaining backward compatibility.

**Comment by apeatling** [comment]
This fixes #20583

File: packages/block-editor/src/autocompleters/block.js (hunk 2 above)
Comment (gziolo) [reply packages/block-editor/src/autocompleters/block.js:21]: Yes, I agree that it is confusing. If you have some ideas on how to improve that, go for it :)

I thought about using so `innerBlocksTemplate` (I called the variable like that in the mapping helper) inside the variation item object but I left what was there before.

**Comment by gziolo** [comment]
I tested a few cases and it works with both 3rd party block variations and with variations registered for core blocks:
Screen Shot 2020-06-30 at 16 11 01
Screen Shot 2020-06-30 at 16 21 34

The only issue I found out is that I can't click to select an item from the dropdown menu. I can only select one of the variations using arrow keys plus `enter`.

**Comment by youknowriad** [comment]
> The only issue I found out is that I can't click to select an item from the dropdown menu

We had this issue on master too, I think a rebase should fix it.

**Review by gziolo — APPROVED** [review]
I tested after rebase and everything works perfectly fine. Great work 💯

**Comment by gziolo** [comment]
I quickly drafted https://github.com/WordPress/gutenberg/pull/23585 to see if that would fit as an e2e test covering this integration.

**Comment by mtias** [comment]
Thanks! This really helps tie the inserter APIs together.
